main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```diff
@@ -224,18 +225,24 @@ object Instances {
     * Checks that there is an instance for each super trait of the trait of `inst`,
     * and that the constraints on `inst` entail the constraints on the super instance.
     */
-  private def checkSuperInstances(inst: TypedAst.Instance, root: TypedAst.Root)(implicit sctx: SharedContext, flix: Flix): Unit = inst match {
-    case TypedAst.Instance(_, _, _, trt, tpe, tconstrs, _, _, _, _) =>
+  private def checkSuperInstances(inst: TypedAst.Instance, root: TypedAst.Root, eqEnv: EqualityEnv)(implicit sctx: SharedContext, flix: Flix): Unit = inst match {
+    case TypedAst.Instance(_, _, _, trt, tpe, tconstrs, econstrs, _, _, _, _) =>
+      // Rigidify sub-instance constraint vars in the substitution to ensure that they appear as-written in compiler errors
+      val evars = econstrs.flatMap(econstr => econstr.tpe1.typeVars ++ econstr.tpe2.typeVars).map(_.sym)
+      val tvars = tconstrs.flatMap(tconstr => tconstr.arg.typeVars).map(_.sym)
+      val rigidityEnv = RigidityEnv.ofRigidVars(evars ::: tvars)
```

Comment:
I think creating this rigidityenv should be in the logic of `getInstanceForType` instead of here.